contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,357 @@
+        if (_depositUnderlayingAmount > 0) {
+            require(
+                IERC20(_dealData.underlyingDealToken).balanceOf(dealCreator) >= _depositUnderlayingAmount,
+                "not enough balance"
+            );
+            IERC20(_dealData.underlyingDealToken).transferFrom(dealCreator, address(this), _depositUnderlayingAmount);
+            uint256 currentDealTokenTotal = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+            if (currentDealTokenTotal >= _dealData.underlyingDealTokenTotal) {
+                underlyingDepositComplete = true;
+                purchaseExpiry = block.timestamp + dealData.purchaseDuration;
+                emit DealFullyFunded(address(this), block.timestamp, purchaseExpiry);
+            }
+
+            emit DepositDealToken(_dealData.underlyingDealToken, dealCreator, currentDealTokenTotal);
+        }
+    }
+
+    /**
+     * @dev method for holder to deposit underlying deal tokens
+     * all underlying deal tokens must be deposited for the purchasing period to start
+     * if tokens were deposited directly, this method must still be called to start the purchasing period
+     */
+    function depositUnderlyingTokens(uint256 _depositUnderlyingAmount) public onlyHolder {
+        UpFrontDeal memory _dealData = dealData;
+
+        require(
+            IERC20(_dealData.underlyingDealToken).balanceOf(msg.sender) >= _depositUnderlyingAmount,
+            "not enough balance"
+        );
+        require(!underlyingDepositComplete, "already deposited the total");
+
+        uint256 balanceBeforeTransfer = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        IERC20(_dealData.underlyingDealToken).transferFrom(msg.sender, address(this), _depositUnderlyingAmount);
+        uint256 balanceAfterTransfer = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        uint256 underlyingDealTokenAmount = balanceAfterTransfer - balanceBeforeTransfer;
+
+        if (balanceAfterTransfer >= _dealData.underlyingDealTokenTotal) {
+            underlyingDepositComplete = true;
+            purchaseExpiry = block.timestamp + _dealData.purchaseDuration;
+            emit DealFullyFunded(address(this), block.timestamp, purchaseExpiry);
+        }
+
+        emit DepositDealToken(_dealData.underlyingDealToken, msg.sender, underlyingDealTokenAmount);
+    }
+
+    /**
+     * @dev allows holder to withdraw any excess underlying deal tokens deposited to the contract
+     */
+    function withdrawExcess() external onlyHolder {
+        UpFrontDeal memory _dealData = dealData;
+        uint256 currentBalance = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        require(currentBalance > _dealData.underlyingDealTokenTotal, "no excess to withdraw");
+
+        uint256 excessAmount = currentBalance - _dealData.underlyingDealTokenTotal;
+        IERC20(_dealData.underlyingDealToken).transferFrom(address(this), msg.sender, excessAmount);
```

Comment:
we can just do `transfer` instead of `transferFrom`.

`IERC20(_dealData.underlyingDealToken).transfer(msg.sender, excessAmount);`

Reply:
done